File: java-code/src/main/java/refactoring/conditional_logic/decompose_conditional/before/ChargeCalculator.java
package refactoring.conditional_logic.decompose_conditional.before;

import java.time.LocalDate;

public class ChargeCalculator {

  public static double calculateCharge(LocalDate aDate, double quantity, Plan plan) {
    double charge;

    if (!aDate.isBefore(plan.getSummerStart()) && !aDate.isAfter(plan.getSummerEnd())) {
      charge = quantity * plan.getSummerRate();
    } else {
      charge = quantity * plan.getRegularRate() + plan.getRegularServiceCharge();
    }

    return charge;
  }
}

File: go-code/conditional_logic/decompose_conditional/before/charge_calculator.go
package before

import (
	"time"
)

type ChargeCalculator struct{}

func (cc *ChargeCalculator) calculateCharge(aDate time.Time, quantity float64, plan Plan) float64 {
	var charge float64

	if !aDate.Before(plan.SummerStart) && !aDate.After(plan.SummerEnd) {
		charge = quantity * plan.SummerRate
	} else {
		charge = quantity*plan.RegularRate + plan.RegularServiceCharge
	}

	return charge
}
